### PR TITLE
Improved Moment Shadows

### PR DESCRIPTION
**Pull request by anisunity**
Adding a Very High shadow configuration
Implementing the Improved Moment Shadow algorith
Adding the very high test scene

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDShadowManager.cs [hunk 1]
```diff
@@ -350,6 +366,13 @@ public void RenderShadows(ScriptableRenderContext renderContext, CommandBuffer c
             // Clear atlas render targets and draw shadows
             m_Atlas.RenderShadows(renderContext, cmd, dss);
             m_CascadeAtlas.RenderShadows(renderContext, cmd, dss);
+
+            // Alright, if we are in very high quality we need to compute the sum tables for the cascade shadow map
+            var hdAsset = (GraphicsSettings.renderPipelineAsset as HDRenderPipelineAsset);
+            if (hdAsset.renderPipelineSettings.hdShadowInitParams.shadowQuality == HDShadowQuality.VeryHigh)
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDShadowManager.cs]: Something that I don't like is that if someone change very high to be something else, like PCSS, then we will still compute everything as moment shadow, the test shouldn't be "area we very high", the test should be , are we moment shadow.

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDShadowSampling.hlsl [hunk 2]
```diff
@@ -282,3 +282,7 @@ float SampleShadow_PCSS(float3 tcs, float2 posSS, float2 scale, float2 offset, f
     //3) Filter
     return PCSS(tcs, filterSize, scale, offset, sampleBias, sampleJitter, tex, compSamp, filterSampleCount);
 }
+
+#ifdef SHADOW_VERY_HIGH
+#include "Packages/com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDIMS.hlsl"
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDShadowSampling.hlsl]: Same remark here, we shouldn't assume that very high is moment shadow

File: com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs [hunk 3]
```diff
@@ -1316,7 +1316,7 @@ protected override void Render(ScriptableRenderContext renderContext, Camera[] c
                         using (new ProfilingSample(cmd, "Render shadows", CustomSamplerId.RenderShadows.GetSampler()))
                         {
                             // This call overwrites camera properties passed to the shader system.
-                            m_LightLoop.RenderShadows(renderContext, cmd, cullingResults);
+                            m_LightLoop.RenderShadows(renderContext, cmd, cullingResults, hdCamera);
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/RenderPipeline/HDRenderPipeline.cs]: don't forget all the master node, check the .template files

File: com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDIMS.hlsl [hunk 4]
```diff
@@ -0,0 +1,406 @@
+/*
+// HARD Improved moment shadows
+void Compute4MomentUnboundedShadowIntensity(out float OutShadowIntensity, float4 Biased4Moments, float FragmentDepth, float DepthBias)
+{
+    // Use short-hands for the many formulae to come
```
Comment (sebastienlagarde) [inline com.unity.render-pipelines.high-definition/Runtime/Lighting/Shadow/HDIMS.hlsl]: is this files used ? is it the file from the article? any license to include here ?

**Comment by anisunity** [comment]
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=ims&automation-tools_branch=add-platform-filter&unity_branch=trunk

**Pull request merged**